src/cpp/include/openvino/genai/rag/text_embedding_pipeline.hpp

```diff
@@ -0,0 +1,172 @@
+// Copyright (C) 2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <filesystem>
+#include <optional>
+#include <variant>
+
+#include "openvino/genai/tokenizer.hpp"
+
+namespace ov {
+namespace genai {
+
+using EmbeddingResult = std::variant<std::vector<float>, std::vector<int8_t>, std::vector<uint8_t>>;
+using EmbeddingResults =
+    std::variant<std::vector<std::vector<float>>, std::vector<std::vector<int8_t>>, std::vector<std::vector<uint8_t>>>;
+
+class OPENVINO_GENAI_EXPORTS TextEmbeddingPipeline {
+public:
+    enum class PoolingType {
+        /**
+         * @brief First token embeddings
+         */
+        CLS = 0,
+        /**
+         * @brief The average of all token embeddings
+         */
+        MEAN = 1,
+    };
+
+    struct OPENVINO_GENAI_EXPORTS Config {
+        /**
+         * @brief Maximum length of tokens passed to the embedding model
+         */
+        std::optional<size_t> max_length;
+
+        /**
+         * @brief Pooling strategy applied to model output tensor
+         */
+        PoolingType pooling_type = PoolingType::CLS;
+
+        /**
+         * @brief If 'true' normalization is applied to embeddings
```

Comment:
```suggestion
         * @brief If 'true' L2 normalization along axis 1 is applied to embeddings
```
Maybe the member and the Property should be renamed to `normize_l2` as well to make it more verbose.

Reply:
The normalization algorithm is implementation detail. [llama_index](https://github.com/run-llama/llama_index/blob/main/llama-index-integrations/embeddings/llama-index-embeddings-huggingface-optimum-intel/llama_index/embeddings/huggingface_optimum_intel/base.py#L18) and [langchain](https://github.com/langchain-ai/langchain-community/blob/main/libs/community/langchain_community/embeddings/openvino.py#L150) do not specify explicitly what type of normalization would be applied.

Reply:
This implementation detail affects the result so it can't be fully hidden. An example of the detail that can be hidden is sorting algorithm in sort(). And even then it's specified if the order of equal elements is guaranteed to be preserved.

As for llama_index and langchain, you can add a note to the docstring that L2 is aligned with them. You can also submit them PRs adding this to see their opinion.

Reply:
Mean comes before L2, so the last point is fine

Reply:
I added L2 norm note to docstrings. Don't think `along axis 1` is needed as normalization is done after pooling.